lib/rubocop/cop/factory_bot/redundant_factory_option.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module FactoryBot
+      # Remove redundant `factory` option.
+      #
+      # @example
+      #   # bad
+      #   association :user, factory: :user
+      #
+      #   # good
+      #   association :user
+      class RedundantFactoryOption < ::RuboCop::Cop::Base
+        extend AutoCorrector
+
+        include RangeHelp
+
+        MSG = 'Remove redundant `factory` option.'
+
+        RESTRICT_ON_SEND = %i[association].freeze
+
+        # @!method association_name_from(node)
+        def_node_matcher :association_name_from, <<~PATTERN
+          (send
+            nil?
+            _
+            (sym $_)
+            ...
+          )
+        PATTERN
+
+        # @!method factory_option_from(node)
+        def_node_matcher :factory_option_from, <<~PATTERN
+          (send
+            nil?
+            _
+            _
+            (hash
+              <
+                $(pair
+                  (sym :factory)
+                  {
+                    (sym $_)
+                    (array (sym $_))
+                  }
+                )
+                ...
+              >
+            )
+          )
+        PATTERN
+
+        def on_send(node)
+          association_name = association_name_from(node)
+          factory_option, factory_value = factory_option_from(node)
+          return if association_name != factory_value
+
+          add_offense(factory_option) do |corrector|
+            autocorrect(corrector, factory_option)
+          end
+        end
```

Comment:
Those two patterns can be combined.
And this one handles the case with a positional trait argument.
```suggestion
        def_node_matcher :association_with_a_factory_option, <<~PATTERN
          (send nil? :association
            (sym $_association_name)
            ...
            (hash
              <$(pair
                  (sym :factory)
                  {
                    (sym $_factory_name)
                    (array (sym $_factory_name))
                  }
                )
                ...
              >
            )
          )
        PATTERN

        def on_send(node)
          association_with_a_factory_option(node) do |association_name, factory_option, factory_name|
            next if association_name != factory_name

            add_offense(factory_option) do |corrector|
              autocorrect(corrector, factory_option)
            end
          end
        end
```